### ISSUE TITLE
预训练代码中随机种子设置错误将导致所有进程生成相同随机数

### ISSUE THREAD

**Issue by jianliao** [jingyaogong/minimind#318]
你好👋

在用 DeepSeek R1一起学习你的代码的时候，我们发现了一个在分布式训练时潜在的 Bug，具体就在[这一行](https://github.com/jingyaogong/minimind/blob/278ec760a191bbdb55139f67f0d3cd6138c10062/train_pretrain.py#L149)。

**描述：**

在使用 PyTorch 分布式训练进行 LLM 预训练时，发现所有进程均使用相同的随机种子（`torch.manual_seed(1337)`），从而导致所有进程生成完全相同的随机数。这会影响模型权重初始化、dropout 以及其他依赖随机性的操作，可能导致模型缺乏足够的随机性，从而影响收敛速度和最终性能。

**复现步骤：**
1. 代码
```python
import os
import torch
import torch.distributed as dist

def main():
    # 从环境变量中获取 rank、world_size 和 local_rank
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    local_rank = int(os.environ["LOCAL_RANK"])
    
    # 初始化分布式环境
    dist.init_process_group(
        backend="nccl",
        init_method="env://",
        world_size=world_size,
        rank=rank
    )
    torch.cuda.set_device(local_rank)

    # 设置固定种子
    torch.manual_seed(1337)
    
    # 生成随机数
    a = torch.rand(3).to(local_rank)
    b = torch.rand(3).to(local_rank)
    
    print(f"Rank {rank}: a={a.cpu()}, b={b.cpu()}")

if __name__ == "__main__":
    main()
```
 2. 使用如下命令运行脚本（假设脚本名为 `torch_manual_seed_bug.py`）：
```bash
OMP_NUM_THREADS=1 torchrun --nproc_per_node=2 torch_manual_seed_bug.py
```
3. 观察输出，发现所有进程生成的随机数完全一致。

**解决方案：**

小规模重构代码，在设定随机数种子的时候，确保每个进程的种子不相同：
`torch.manual_seed(1337 + rank)`

**Comment by jingyaogong** [comment]
是一个潜在的问题，谢谢你的见解，会更新😊
